sycl/doc/Assert.md

```diff
@@ -208,20 +223,77 @@ struct _pi_device_binary_property_struct {
 };
 ```
 
-Whenever `isAssertEnabled` property set is present, this specific device image
-was built with `NDEBUG` macro undefined and it requires fallback implementation
-of `__devicelib_assert_fail` (i.e. if Device-side Runtime doesn't support it).
+A distinct property set `SYCL/assert used` is added. In this set a property
+with the name of the kernel is added whenever the kernel uses assert. Use of
+assert is detected through call to `__devicelib_assert_fail` function after
+linking device binary image with wrapper device library (the `libsycl-crt`
+library).
+
+The property set and the underlying properties are added by `sycl-post-link`
+tool with help of building callgraph for each and every kernel in device binary
+image.
+
+The added property is used for:
+ - deciding if online-linking against fallback devicelib is required;
+ - if there's a need to enqueue program scope variable copier kernel and checker
+   host-task.
+
+Suppose the following example user code:
+```c++
+void user_func(int X) {
+  assert(X && “X is nil”);
+}
+
+int main() {
+  queue Q(...);
+  Q.submit([&] (handler& CGH) {
+    CGH.single_task([=] () {
+      do_smth();
+      user_func(0);
+      do_smth_else();
+    });
+  });
+  ...
+}
+```
+
+The following LLVM IR pseudo code will be generated after linking against
+fallback implementation of devicelib:
+```
+@AssertHappenedMem = global AssertHappened
+
+/// user's code
+void user_func(int X) {
+if (!(X && “X is nil")) {
+    __assert_fail(...);
+  }
+}
 
-Any properties in `isAssertEnabled` property set are ignored.
+kernel(...) {
+  do_smth()
+  user_func(0);
+  do_smth_else();
+}
 
-The property set is added to device binary descriptor whenever at least single
-translation unit was compiled with assertions enabled i.e. `NDEBUG` undefined.
+/// __assert_fail belongs to Linux version of devicelib
+void __assert_fail(...) {
+  ...
+  __devicelib_assert_fail(...);
+}
 
-The property set is added by `sycl-post-link` tool depending on module metadata.
-Metadata is provided by Clang frontend. Metadata name is `is_assert_enabled`.
+void __devicelib_assert_fail(Expr, File, Line, GlobalID, LocalID) {
+  ...
+  volatile int *Ptr = (volatile int *)AssertHappenedMem.Flag;
+  int Expected = 0;
+  int Desired = 1;
 
+  if (atomic_CAS(&AssertHappenedMem.Flag, Expected, Desired))
```

Comment:
```suggestion
  if (atomic_CAS(AssertHappenedMem.Flag, Expected, Desired))
```
?

Reply:
Atomic operation requires referencing the variable, not copying its value.